From 3273a095cf4e4a3f1594f735c7b9687e3869c898 Mon Sep 17 00:00:00 2001
From: Rob Clark <rclark@redhat.com>
Date: Wed, 17 Dec 2014 21:38:20 -0500
Subject: [drm] radeon/dpm: handle voltage info fetching on hawaii

Message-id: <1418852380-13061-10-git-send-email-rclark@redhat.com>
Patchwork-id: 102151
O-Subject: [RHEL7 drm 09/89] drm/radeon/dpm: handle voltage info fetching on hawaii
Bugzilla: 1173317
RH-Acked-by: Jerome Glisse <jglisse@redhat.com>
RH-Acked-by: Dave Airlie <airlied@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1173317
Upstream: since v3.16.4

commit 88086e07b486d432d5047d12d60f75148c1efeff

Author:     Alex Deucher <alexander.deucher@amd.com>
AuthorDate: Thu Jul 31 18:07:17 2014 -0400
Commit:     Rob Clark <rclark@redhat.com>
CommitDate: Wed Dec 10 17:08:37 2014 -0500

    drm/radeon/dpm: handle voltage info fetching on hawaii

    commit 6b57f20cb5b708415fbab63847f8f8429b051af8 upstream.

    Some hawaii cards use a different method to fetch the
    voltage info from the vbios.

    bug:
    https://bugs.freedesktop.org/show_bug.cgi?id=74250

    Signed-off-by: Alex Deucher <alexander.deucher@amd.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

Signed-off-by: Rob Clark <rclark@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/radeon/ci_dpm.c b/drivers/gpu/drm/radeon/ci_dpm.c
index 584090a..022561e 100644
--- a/drivers/gpu/drm/radeon/ci_dpm.c
+++ b/drivers/gpu/drm/radeon/ci_dpm.c
@@ -940,7 +940,18 @@ static void ci_get_leakage_voltages(struct radeon_device *rdev)
  pi->vddc_leakage.count = 0;
  pi->vddci_leakage.count = 0;
 
- if (radeon_atom_get_leakage_id_from_vbios(rdev, &leakage_id) == 0) {
+ if (rdev->pm.dpm.platform_caps & ATOM_PP_PLATFORM_CAP_EVV) {
+  for (i = 0; i < CISLANDS_MAX_LEAKAGE_COUNT; i++) {
+   virtual_voltage_id = ATOM_VIRTUAL_VOLTAGE_ID0 + i;
+   if (radeon_atom_get_voltage_evv(rdev, virtual_voltage_id, &vddc) != 0)
+    continue;
+   if (vddc != 0 && vddc != virtual_voltage_id) {
+    pi->vddc_leakage.actual_voltage[pi->vddc_leakage.count] = vddc;
+    pi->vddc_leakage.leakage_id[pi->vddc_leakage.count] = virtual_voltage_id;
+    pi->vddc_leakage.count++;
+   }
+  }
+ } else if (radeon_atom_get_leakage_id_from_vbios(rdev, &leakage_id) == 0) {
   for (i = 0; i < CISLANDS_MAX_LEAKAGE_COUNT; i++) {
    virtual_voltage_id = ATOM_VIRTUAL_VOLTAGE_ID0 + i;
    if (radeon_atom_get_leakage_vddc_based_on_leakage_params(rdev, &vddc, &vddci,
-- 
1.7.1